### ISSUE TITLE
Add doctest utility to handle deprecation and future warnings

### ISSUE THREAD

**Issue by tobiasdiez** [sagemath/sage#39211]
Currently, we just print out the deprecation warning. However, the output depends on the order of the tests as only the first deprecation warning is actually printed - further warnings are automatically suppressed. Thus, if you run only a single test in a file then this may fail due to now printing a deprecation warning.

Instead, I propose to use an approach similar to sympy which have an utility to check that a deprecation warning is printed - but without actually printing the warning:
- https://docs.sympy.org/latest/modules/testing/pytest.html#sympy.testing.pytest.warns_deprecated_sympy
- https://docs.sympy.org/latest/contributing/deprecations.html#how-to-deprecate-code
- https://docs.sympy.org/latest/modules/testing/runtests.html#sympy.testing.runtests.raise_on_deprecated
- https://github.com/sympy/sympy/blob/57476ba48733061919d7edec6e95c95e0aaa4ee3/sympy/testing/pytest.py#L328

This would also help with pytest compatibility, where all tests are run in the same context and thus may print even fewer warnings.

**Comment by PredictiveManish** [comment]
Can i start working on this? 
Can you give a little guide on how to see the current feature for upgradation as you mentioned for the other deprecation warning from first ones, it will give better idea as I am new to this community and exploring and finding to solve issues.

**Comment by tobiasdiez** [comment]
@PredictiveManish Thanks for your interest and welcome to sage!

What I had in mind is to rewrite, for example, the following example
https://github.com/sagemath/sage/blob/c9dd1e85118c875d12bd7bc66d0817f907cf4e33/src/sage/algebras/algebra.py#L30-L36
as follows
```python
sage: from sage.algebras.algebra import is_Algebra
sage: from sage.doctest import warns_deprecated
sage: R.<x,y> = FreeAlgebra(QQ,2)
sage: with warns_deprecated():
...       is_Algebra(R)
True
```

So the method `warns_deprecated` collects the warning, and internally checks that a `DeprecationWarning` is raised. But it is not displayed.

The inconsistent behavior can be seen e.g. in
https://github.com/sagemath/sage/blob/c9dd1e85118c875d12bd7bc66d0817f907cf4e33/src/sage/algebras/lie_algebras/free_lie_algebra.py#L55-L63
and https://github.com/sagemath/sage/blob/c9dd1e85118c875d12bd7bc66d0817f907cf4e33/src/sage/algebras/lie_algebras/free_lie_algebra.py#L77-L79

The code is the same, but in the first example a future warning is printed but not in the second. This only works if the first code is executed before the second one.

**Comment by user202729** [comment]
The problem with `with warns_deprecated` is that it isn't as clear to the documentation reader that what the desired behavior should be. Previously it is reasonably clear that it's intended the warnings is printed out.

What about just tell pytest to reset warning status (whatever it is) at the start of each file?

**Comment by tobiasdiez** [comment]
> The problem with `with warns_deprecated` is that it isn't as clear to the documentation reader that what the desired behavior should be. Previously it is reasonably clear that it's intended the warnings is printed out.

I agree, this is a disadvantage. On the other hand, the deprecation should be documented in the main description of the method anyway.

> What about just tell pytest to reset warning status (whatever it is) at the start of each file?

Pytest doesn't really have a notion of "file" - once the tests are collected, it's only a bunch of tests. Moreover, you may also want to run single tests independently (e.g. if the second test above in `free_lie_algebra` fails you only want to run this one) or in random order (https://doc.sagemath.org/html/en/developer/doctesting.html#running-tests-in-a-random-order)

**Comment by user202729** [comment]
Then what about telling both the current monolithic test system and pytest to reset warning status on each docstring? Same result.

(might need a lot of fixes though. At least we have sage-fixdoctests which works somewhat.)

**Comment by tobiasdiez** [comment]
I guess it comes down what you want to achieve with these deprecation warnings in tests:
- A: Raise awareness that this is deprecated/unstable code; in this case it would be indeed the best to show it on every test that uses this code
- B: Just test that (under certain conditions) a deprecation warning is shown; in this case, normally don't print a deprecation warning at all and use the `warns_deprecated` helper once.

I'm more in camp B.

What do you think about the following: do not show the deprecation warning normally, but in the `warns_deprecated` context it will print out the full deprecation warning as currently? Then it would be very similar to the current display, at the only expense that the doctest now has to invoke `warns_deprecated`.

**Comment by user202729** [comment]
no strong opinion, but side note:

> at the only expense that the doctest now has to invoke warns_deprecated.

depends on what you mean.

1. You want every doctest to have this: might as well put it in forker/test invoker code
2. you want only some doctest to have this: might as well make doctest parser parse some magic comment like `# show all warnings`
